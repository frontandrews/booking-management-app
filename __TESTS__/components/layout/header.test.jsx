import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { useRouter, usePathname } from 'next/navigation';
import Header from '@/components/layout/header';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn(),
}));

jest.mock('@/redux/features/auth/slice', () => ({
  logOut: jest.fn(),
}));

jest.mock('@/components/layout/desktop-navigation', () => ({
  DesktopNavigation: () => (
    <div data-testid="desktop-navigation">DesktopNavigation</div>
  ),
}));

jest.mock('@/components/layout/mobile-navigation', () => ({
  MobileNavigation: ({ mobileMenuOpen, setMobileMenuOpen }) => (
    <div
      data-testid="mobile-navigation"
      onClick={() => setMobileMenuOpen(!mobileMenuOpen)}
    >
      MobileNavigation {mobileMenuOpen ? 'Open' : 'Closed'}
    </div>
  ),
}));

const createMockStore = (initialState) => {
  return configureStore({
    reducer: {
      auth: () => initialState.auth,
    },
    preloadedState: initialState,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  });
};

describe('Header Component', () => {
  let store;
  const mockPush = jest.fn();
  const mockPathname = '/';

  const initialState = {
    auth: {
      user: {
        isAuthenticated: false,
      },
    },
  };

  beforeEach(() => {
    store = createMockStore(initialState);
    useRouter.mockReturnValue({ push: mockPush });
    usePathname.mockReturnValue(mockPathname);
  });

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <Header />
      </Provider>,
    );

  test('renders Header correctly', () => {
    renderComponent();

    expect(screen.getByText('HostSoft')).toBeInTheDocument();
    expect(screen.getByTestId('desktop-navigation')).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /open main menu/i }),
    ).toBeInTheDocument();
  });

  test('renders login link when not authenticated', () => {
    renderComponent();

    expect(screen.getByText('Log in')).toBeInTheDocument();
  });

  test('renders logout link when authenticated', () => {
    store = createMockStore({
      auth: {
        user: {
          isAuthenticated: true,
        },
      },
    });

    renderComponent();

    expect(screen.getByText('Logout')).toBeInTheDocument();
  });

  test.todo('handles logout click');

  test('handles mobile menu toggle', () => {
    renderComponent();

    const openMenuButton = screen.getByRole('button', {
      name: /open main menu/i,
    });
    fireEvent.click(openMenuButton);

    expect(screen.getByTestId('mobile-navigation')).toHaveTextContent(
      'MobileNavigation Open',
    );

    const mobileNavigation = screen.getByTestId('mobile-navigation');
    fireEvent.click(mobileNavigation);

    expect(screen.getByTestId('mobile-navigation')).toHaveTextContent(
      'MobileNavigation Closed',
    );
  });
});
