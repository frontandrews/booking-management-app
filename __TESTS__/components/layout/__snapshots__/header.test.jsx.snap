// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component handles mobile menu toggle 1`] = `
<div>
  <header
    class="bg-primary"
  >
    <nav
      aria-label="Global"
      class="flex items-center justify-between p-6 lg:px-8 container mx-auto"
    >
      <div
        class="flex lg:flex-1"
      >
        <a
          class="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          href="/"
        >
          HostSoft
        </a>
      </div>
      <div
        class="flex lg:hidden"
      >
        <button
          class="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-400"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        data-testid="desktop-navigation"
      >
        DesktopNavigation
      </div>
      <div
        class="hidden lg:flex lg:flex-1 lg:justify-end"
      >
        <a
          class="text-sm font-semibold leading-6 text-white"
          href="/sign-in"
        >
          Log in 
          <span
            aria-hidden="true"
          >
            →
          </span>
        </a>
      </div>
    </nav>
    <div
      data-testid="mobile-navigation"
    >
      MobileNavigation 
      Closed
    </div>
  </header>
</div>
`;

exports[`Header Component renders Header correctly 1`] = `
<div>
  <header
    class="bg-primary"
  >
    <nav
      aria-label="Global"
      class="flex items-center justify-between p-6 lg:px-8 container mx-auto"
    >
      <div
        class="flex lg:flex-1"
      >
        <a
          class="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          href="/"
        >
          HostSoft
        </a>
      </div>
      <div
        class="flex lg:hidden"
      >
        <button
          class="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-400"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        data-testid="desktop-navigation"
      >
        DesktopNavigation
      </div>
      <div
        class="hidden lg:flex lg:flex-1 lg:justify-end"
      >
        <a
          class="text-sm font-semibold leading-6 text-white"
          href="/sign-in"
        >
          Log in 
          <span
            aria-hidden="true"
          >
            →
          </span>
        </a>
      </div>
    </nav>
    <div
      data-testid="mobile-navigation"
    >
      MobileNavigation 
      Closed
    </div>
  </header>
</div>
`;

exports[`Header Component renders login link when not authenticated 1`] = `
<div>
  <header
    class="bg-primary"
  >
    <nav
      aria-label="Global"
      class="flex items-center justify-between p-6 lg:px-8 container mx-auto"
    >
      <div
        class="flex lg:flex-1"
      >
        <a
          class="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          href="/"
        >
          HostSoft
        </a>
      </div>
      <div
        class="flex lg:hidden"
      >
        <button
          class="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-400"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        data-testid="desktop-navigation"
      >
        DesktopNavigation
      </div>
      <div
        class="hidden lg:flex lg:flex-1 lg:justify-end"
      >
        <a
          class="text-sm font-semibold leading-6 text-white"
          href="/sign-in"
        >
          Log in 
          <span
            aria-hidden="true"
          >
            →
          </span>
        </a>
      </div>
    </nav>
    <div
      data-testid="mobile-navigation"
    >
      MobileNavigation 
      Closed
    </div>
  </header>
</div>
`;

exports[`Header Component renders logout link when authenticated 1`] = `
<div>
  <header
    class="bg-primary"
  >
    <nav
      aria-label="Global"
      class="flex items-center justify-between p-6 lg:px-8 container mx-auto"
    >
      <div
        class="flex lg:flex-1"
      >
        <a
          class="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          href="/bookings"
        >
          HostSoft
        </a>
      </div>
      <div
        class="flex lg:hidden"
      >
        <button
          class="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-400"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        data-testid="desktop-navigation"
      >
        DesktopNavigation
      </div>
      <div
        class="hidden lg:flex lg:flex-1 lg:justify-end"
      >
        <div
          class="text-sm font-semibold leading-6 text-white cursor-pointer"
        >
          Logout 
          <span
            aria-hidden="true"
          >
            →
          </span>
        </div>
      </div>
    </nav>
    <div
      data-testid="mobile-navigation"
    >
      MobileNavigation 
      Closed
    </div>
  </header>
</div>
`;
